File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.cpp
class Solution {
public:
    int longestOnes(vector<int>& nums, int k) {
        

        int n=nums.size();

        int l=0; 
        int r=0;
        int zeroCount=0;
        int maxLen=0;
        while(r<n){
            zeroCount+= nums[r]==0?1:0;
            //handling invalid condition
            while(zeroCount>k){
                zeroCount-= nums[l]==0? 1:0;
                l++;
            }
            maxLen=max(maxLen, r-l+1);
            r++;
        }
        return maxLen;
    }
};

File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.java
class Solution {
    public int longestOnes(int[] nums, int k) {
        int n=nums.length;

        int l=0;
        int r=0;

        int zero=0;
        int ans=0;

        while(r<n){
            zero += nums[r]==0 ? 1:0;
            while(zero>k){
                zero-= nums[l]==1?0 : 1;
                l++;
            }
            ans=Math.max(ans, r-l+1);
            r++;
        }

        return ans;
    }
}